### ISSUE TITLE
Issue scraping through https

### ISSUE THREAD

**Issue by teddb** [scrapy/scrapy#1434]
I can successfully scrape an http page through a proxy. I can also scrape an https page without using a proxy. I cannot scrape an https page through a proxy.

My spider works successfully if the 's' in 'https' of the target page is removed or I disable the proxy. I can access the https page through the proxy through my browser.

Here is what I added to the settings file:

    DOWNLOADER_MIDDLEWARES = {'scrapy.downloadermiddlewares.httpproxy.HttpProxyMiddleware': 110, 'test_website.middlewares.ProxyMiddleware': 100}
    PROXIES = [{'ip_port': 'us-il.proxymesh.com:31280', 'user_pass': 'username:password'}]

Here is the code for my spider:

    import scrapy
     
    class TestSpider(scrapy.Spider):
        name = "test_spider"
        allowed_domains = "ipify.org"
         start_urls = ["https://api.ipify.org"]
     
        def parse(self, response):
            with open('test.html', 'wb') as f:
                f.write(response.body)

Here is the middlewares file:

    import base64
    import random
    from settings import PROXIES
     
    class ProxyMiddleware(object):
        def process_request(self, request, spider):
            proxy = random.choice(PROXIES)
            if proxy['user_pass'] is not None:
                request.meta['proxy'] = "http://%s" % proxy['ip_port']
                encoded_user_pass = base64.encodestring(proxy['user_pass'])
                request.headers['Proxy-Authorization'] = 'Basic ' + encoded_user_pass            
            else:
                request.meta['proxy'] = "http://%s" % proxy['ip_port']

Here is the log file:

        2015-08-12 20:15:50 [scrapy] INFO: Scrapy 1.0.3 started (bot: test_website)
    2015-08-12 20:15:50 [scrapy] INFO: Optional features available: ssl, http11
    2015-08-12 20:15:50 [scrapy] INFO: Overridden settings: {'NEWSPIDER_MODULE': 'test_website.spiders', 'SPIDER_MODULES': ['test_website.spiders'], 'LOG_STDOUT': True, 'LOG_FILE': 'log.txt', 'BOT_NAME': 'test_website'}
    2015-08-12 20:15:51 [scrapy] INFO: Enabled extensions: CloseSpider, TelnetConsole, LogStats, CoreStats, SpiderState
    2015-08-12 20:15:53 [scrapy] INFO: Enabled downloader middlewares: ProxyMiddleware, HttpAuthMiddleware, DownloadTimeoutMiddleware, UserAgentMiddleware, RetryMiddleware, DefaultHeadersMiddleware, MetaRefreshMiddleware, HttpCompressionMiddleware, RedirectMiddleware, CookiesMiddleware, ChunkedTransferMiddleware, DownloaderStats
    2015-08-12 20:15:53 [scrapy] INFO: Enabled spider middlewares: HttpErrorMiddleware, OffsiteMiddleware, RefererMiddleware, UrlLengthMiddleware, DepthMiddleware
    2015-08-12 20:15:53 [scrapy] INFO: Enabled item pipelines: 
    2015-08-12 20:15:53 [scrapy] INFO: Spider opened
    2015-08-12 20:15:53 [scrapy] INFO: Crawled 0 pages (at 0 pages/min), scraped 0 items (at 0 items/min)
    2015-08-12 20:15:53 [scrapy] DEBUG: Telnet console listening on 127.0.0.1:6023
    2015-08-12 20:15:53 [scrapy] DEBUG: Retrying <GET https://api.ipify.org> (failed 1 times): [<twisted.python.failure.Failure <class 'OpenSSL.SSL.Error'>>]
    2015-08-12 20:15:53 [scrapy] DEBUG: Retrying <GET https://api.ipify.org> (failed 2 times): [<twisted.python.failure.Failure <class 'OpenSSL.SSL.Error'>>]
    2015-08-12 20:15:53 [scrapy] DEBUG: Gave up retrying <GET https://api.ipify.org> (failed 3 times): [<twisted.python.failure.Failure <class 'OpenSSL.SSL.Error'>>]
    2015-08-12 20:15:53 [scrapy] ERROR: Error downloading <GET https://api.ipify.org>: [<twisted.python.failure.Failure <class 'OpenSSL.SSL.Error'>>]
    2015-08-12 20:15:53 [scrapy] INFO: Closing spider (finished)
    2015-08-12 20:15:53 [scrapy] INFO: Dumping Scrapy stats:
    {'downloader/exception_count': 3,
     'downloader/exception_type_count/twisted.web._newclient.ResponseNeverReceived': 3,
     'downloader/request_bytes': 819,
     'downloader/request_count': 3,
     'downloader/request_method_count/GET': 3,
     'finish_reason': 'finished',
     'finish_time': datetime.datetime(2015, 8, 13, 2, 15, 53, 943000),
     'log_count/DEBUG': 4,
     'log_count/ERROR': 1,
     'log_count/INFO': 7,
     'scheduler/dequeued': 3,
     'scheduler/dequeued/memory': 3,
     'scheduler/enqueued': 3,
     'scheduler/enqueued/memory': 3,
     'start_time': datetime.datetime(2015, 8, 13, 2, 15, 53, 38000)}
    2015-08-12 20:15:53 [scrapy] INFO: Spider closed (finished)

**Comment by ianzhang001** [comment]
@teddb have you resolved this issues? I meet the same problem.

**Comment by ianzhang001** [comment]
@dangra i meet the same problem when i use proxy and crawl https site

**Comment by ngtrieuvi92** [comment]
@teddb  i meet the sample issue?  have you resolved this?

**Comment by DixonShen** [comment]
maybe the proxy u use doesn't support the https protocol

**Comment by wmadaus** [comment]
I had issues with the same thing. @DixonShen is right, it is probably and issue with the proxy not supporting it. Search for ones that support HTTPS. [https://www.us-proxy.org/](https://www.us-proxy.org/,) you can sort by HTTPS support.

**Comment by yssource** [comment]
The same problem happens to me as well.
I am able to curl https website by CURL
`curl -x http://ip:port -L https://www.google.com`
But, it will get error "Could not open CONNECT tunnel with proxy", when scrapying https by proxy.

**Comment by kmike** [comment]
Hey @yssource,

What Scrapy version are you using? If it is Scrapy 1.1, could you please paste output of `scrapy version -v`?

**Comment by yssource** [comment]
@kmike 
```
Scrapy    : 1.1.0
lxml      : 3.6.0.0
libxml2   : 2.9.3
Twisted   : 16.2.0
Python    : 3.4.3 (default, Mar 31 2016, 20:42:37) - [GCC 5.3.1 20151207 (Red Hat 5.3.1-2)]
pyOpenSSL : 16.0.0 (OpenSSL 1.0.2h-fips  3 May 2016)
Platform  : Linux-4.4.9-300.fc23.x86_64-x86_64-with-fedora-23-Twenty_Three
```
my settings looks like

```
DOWNLOADER_MIDDLEWARES = {
    'misc.middleware.CustomHttpProxyMiddleware': 400,
    'misc.middleware.CustomUserAgentMiddleware': 401,
    # 'scrapy.downloadermiddlewares.httpproxy.HttpProxyMiddleware': 410,
}
```
```
class CustomHttpProxyMiddleware(object):

    def process_request(self, request, spider):
        # TODO implement complex proxy providing algorithm
        if self.use_proxy(request):
            p = random.choice(PROXIES)
            try:
                request.meta['proxy'] = "http://%s" % p['ip_port']
            except Exception as e:
                log.critical("Exception %s" % e)

    def use_proxy(self, request):
        return True
```

There is no problem for http only website with proxy settings.
For instance, http://www.google.com is Ok. But, https://www.google.com, reporting "Could not open CONNECT tunnel with proxy" .
If disable "misc.middleware.CustomHttpProxyMiddleware", both http and https work well.

And  curl works well as well 
```
curl -x http://ip:port -L https://www.google.com
curl -x http://ip:port -L http://www.google.com
```

**Comment by Lachoneous** [comment]
@kmike 

We are having a similar problem in Scrapy when using proxies to target https urls. "Could not open connect tunnel with proxy."  This includes proxies that are supposed to support https.  Curious if there are  any insights into this.  We are using Scrapy 1.1RC4.

**Comment by redapple** [comment]
@yssource , my local tests with HTTPS proxies from http://proxylist.hidemyass.com/ work well with:
```
$ scrapy version -v
Scrapy    : 1.1.0
lxml      : 3.6.0.0
libxml2   : 2.9.3
Twisted   : 16.2.0
Python    : 3.5.1+ (default, Mar 30 2016, 22:46:26) - [GCC 5.3.1 20160330]
pyOpenSSL : 16.0.0 (OpenSSL 1.0.2g-fips  1 Mar 2016)
Platform  : Linux-4.4.0-24-generic-x86_64-with-Ubuntu-16.04-xenial
```
and I've checked with Wireshark that the proxy is indeed contacted.

If you still have a failing proxy+https website example, you can send it to me by email (address on my github card) and I can try reproducing

**Comment by redapple** [comment]
@Lachoneous , as [I commented above](https://github.com/scrapy/scrapy/issues/1434#issuecomment-227098583), if you have an example failing proxy+website, you can send it to me to reproduce

**Comment by kalyanp** [comment]
@redapple , I have sent an email to you describing the problem, sample python code and proxies to reproduce the problem.

**Comment by kalyanp** [comment]
@redapple , Forgot to mention. I am replying on behalf of @Lachoneous

**Comment by redapple** [comment]
@kalyanp , thanks for the email.
it's related to missing `Host: somehost...` header in the CONNECT request that Scrapy's tunneling agent sends.

Compare curl's
```
CONNECT httpbin.org:443 HTTP/1.1
Host: httpbin.org:443
User-Agent: curl/7.47.0
Proxy-Connection: Keep-Alive

HTTP/1.0 200 Connection Established
Proxy-agent: Apache
...
```
with Scrapy's

```
CONNECT httpbin.org:443 HTTP/1.1

HTTP/1.1 400 Bad Request
Date: Tue, 21 Jun 2016 08:27:52 GMT
Server: Apache/2.2.15 (CentOS)
Content-Length: 304
Connection: close
Content-Type: text/html; charset=iso-8859-1

<!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML 2.0//EN">
<html><head>
<title>400 Bad Request</title>
</head><body>
<h1>Bad Request</h1>
<p>Your browser sent a request that this server could not understand.<br />
</p>
<hr>
<address>Apache/2.2.15 (CentOS) Server at httpbin.org Port 443</address>
</body></html>
```

`Host:` header is a MUST according to [RFC 2616](https://tools.ietf.org/html/rfc2616#page-128)

I'm preparing a PR to fix this (tested it locally, worked fine)

**Comment by redapple** [comment]
@kalyanp , @Lachoneous ,
are you able to test https://github.com/scrapy/scrapy/issues/2069  to check if it fixes your trouble on your end?

**Comment by kalyanp** [comment]
@redapple , I pulled the changes you made and I was doing some testing. So far I haven't seen any issues and that change actually fixed the issue. Thanks for the quick response and the fix. Now I see clear messages in the logs too which is very helpful.

**Comment by Lachoneous** [comment]
Thanks @redapple !  Your quick response to this issue is nothing but awesome.  We'll keep testing and let you know if we find any other issues, but so far this fix appears to resolve the CONNECT issue in a remarkable way.

**Comment by redapple** [comment]
https://github.com/scrapy/scrapy/pull/2069 is merged and will be available in Scrapy 1.1.1
